pfcpiface/aether.go

```diff
@@ -0,0 +1,849 @@
+	c := exec.CommandContext(ctx, "ping", "-c", "1", dst.String())
+	var stdout, stderr bytes.Buffer
+	c.Stdout = &stdout
+	c.Stderr = &stderr
+	err := c.Run()
+	if err != nil {
+		log.Errorln(err, stdout.String(), stderr.String())
+		return err
+	}
+
+	return nil
+}
+
+func (a *aether) resolveNeighbor(ctx context.Context, linkIndex int, ip net.IP) (net.HardwareAddr, error) {
+	// Trigger ARP lookup for gateway IP.
+	if err := ping(ctx, ip); err != nil {
+		log.Errorf("Ping to %v failed: %v", ip, err)
+		return nil, err
+	}
+
+	neighs, err := netlink.NeighList(linkIndex, unix.AF_INET)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, n := range neighs {
+		log.Tracef("%+v", n)
+		if n.State == netlink.NUD_FAILED || n.State == netlink.NUD_NOARP {
+			continue
+		}
+		if n.IP.Equal(ip) {
+			return n.HardwareAddr, nil
+		}
+	}
+
+	return nil, ErrNotFound("neighbor " + ip.String())
+}
+
+func (a *aether) setupRoutingRules() (err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), Timeout)
+	defer cancel()
+
+	// TODO(max): this should be a bridging-like entry
+	if err = a.addIPLookupRule(ctx, a.routeToFabric.Dst, a.gatewayMAC); err != nil {
+		return
+	}
+
+	// Default route over Fabric gateway for encaped uplink traffic.
+	defaultRoute := &net.IPNet{
+		IP:   net.IPv4zero,
+		Mask: net.CIDRMask(0, net.IPv4len*8),
+	}
+	if err = a.addIPLookupRule(ctx, defaultRoute, a.gatewayMAC); err != nil {
+		return
+	}
+
+	return
+}
+
+func (a *aether) linkBessModules(ctx context.Context, m1 string, ogate uint64, m2 string, igate uint64) (err error) {
+	req := &pb.ConnectModulesRequest{
+		M1:               m1,
+		M2:               m2,
+		Ogate:            ogate,
+		Igate:            igate,
+		SkipDefaultHooks: false,
+	}
+
+	if _, err = a.client.ConnectModules(ctx, req); err != nil {
+		log.Errorf("Could not link modules with request %+v: %v", req, err)
+		return
+	}
+
+	return
+}
+
+func (a *aether) findNextFreeOgate(ctx context.Context, module string) (ogate uint64, err error) {
+	req := &pb.GetModuleInfoRequest{Name: module}
+	resp, err := a.client.GetModuleInfo(ctx, req)
+	if err != nil {
+		return
+	}
+
+outer:
+	for i := uint64(0); i < maxGates; i++ {
+		for _, og := range resp.Ogates {
+			if og.Ogate == i {
+				continue outer
+			}
+		}
+		// No collision found, this ogate is free.
+		return i, nil
+	}
+
+	return 0, ErrNotFound("free ogate")
+}
+
+func (a *aether) addIPLookupRule(ctx context.Context, dst *net.IPNet, nhopMAC net.HardwareAddr) (err error) {
+	ones, zeros := dst.Mask.Size()
+	if ones+zeros == 0 {
+		return ErrInvalidArgumentWithReason("addIPLookupRule", dst, "not a CIDR mask")
+	}
+
+	// Create next hop dst MAC packet data update module and add to cache, if needed.
+	n, exists := a.neighborCache[dst.String()]
+	log.Tracef("neighbor cache for %v exists: %v, %+v", dst, exists, n)
+	if !exists {
+		n.nhopMAC = nhopMAC
+		n.updateModuleName, err = a.createDstMacUpdateModule(ctx, dst, nhopMAC)
```

Comment:
Do we need to create a BESS Module at runtime? I lack enough expertise on the BESS side, but creating a new module seems to me a bit heavy (time-consuming?) operation. A pointer why we need to do that and why we cannot use a "static" BESS module would be useful to better understand this part of Aether pipeline.